### PR TITLE
LazyString POC

### PR DESCRIPTION
**Pull request by fabiokr**
Decryption operations are usually slow. If you have a model with a
couple encrypted fields, loading the record will forcefully trigger
decryption for each one of the fields, even if they are not actually
used.

These changes introduce a LazyString object that is returned by the
providers `decrypt` method. The LazyString is a wrapper around the
encrypted String object, and will only trigger decryption if the string
is actually called upon.

```
# the Rails console prints the model attributes on the console, 
# which would trigger the decryption, to avoid that 
# we chain with `"hey"` below

(development) > (b = MyModel.last) && "hey" 
  MyModel Load (0.9ms)  SELECT  "my_models".* FROM "my_models" ORDER BY "my_models"."id" DESC LIMIT $1  [["LIMIT", 1]]
=> "hey"
(development) > b.sensitive_login
   (0.3ms)  BEGIN
   (0.3ms)  SELECT pgp_key_id([FILTERED])
   (0.2ms)  COMMIT
   (0.7ms)  SELECT pgp_sym_decrypt([FILTERED])
=> "theusername"
(development) > b.sensitive_login
=> "theusername" 
(development) > b.sensitive_password
   (0.4ms)  BEGIN
   (0.4ms)  SELECT pgp_key_id([FILTERED])
   (0.2ms)  COMMIT
   (0.7ms)  SELECT pgp_sym_decrypt([FILTERED])
=> "thepassword"
(development) > b.sensitive_password
=> "thepassword"
```

### CODE REVIEW

File: lib/crypt_keeper/provider/base.rb [hunk 1]
```diff
@@ -16,6 +16,34 @@ def load(value)
...
+      # Public: Subclasses should implement their own `encrypt` method.
+      #
+      # value - the String to encrypt
+      #
+      # Returns a String.
+      def encrypt(value)
+        NotImplementedError
+      end
+
+      # Public: Decrypts a string. This will not immediatelly decrypt the String,
```
Comment (itspriddle) [inline lib/crypt_keeper/provider/base.rb]: s/immediatelly/immediately/

**Comment by fabiokr** [comment]
@rubemz what would you expect to see on the Readme? This is supposed to happen on the background, no need to change anything on the clients, it sounds like an implementation detail.

File: lib/crypt_keeper/lazy_string.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    # Public: Implements to_str for String compatibility.
+    #
+    # Returns the decrypted string.
+    def to_str
+      decrypted_string
+    end
+
+    # Public: Implements to_ary for String compatibility.
+    #
+    # Returns the decrypted string.
```
Comment (rubemz) [inline lib/crypt_keeper/lazy_string.rb]: it returns nil, not the decrypted string

Comment (rubemz) [reply lib/crypt_keeper/lazy_string.rb]: btw, why to you need to implement this method and not just handle it in the method_missing

**Comment by rubemz** [comment]
@fabiokr hmm true. No README updated then :D

File: lib/crypt_keeper/lazy_string.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
+module CryptKeeper
+  # LazyString is a wrapper around a String instance to avoid immediatelly
```
Comment (boone) [inline lib/crypt_keeper/lazy_string.rb]: immediatelly typo here too

File: lib/crypt_keeper/lazy_string.rb (hunk 2 above)
Comment (fabiokr) [reply lib/crypt_keeper/lazy_string.rb]: @rubemz Rspec expects this method for some reason, I added it for now to make it shut up, but I'll investigate more.

**Comment by fabiokr** [comment]
Ruby classes implemented in C that expect a real String don't play well with the LazyString. Made method_missing respond to anything, to go around that, but that could also cause other weird behavior.

**Comment by fabiokr** [comment]
@jmazzi For reference on the latest findings: this PR is totally not needed. The Rails serializers are already lazy loaded. The CryptKeeper serializer will only decrypt on demand. The behavior we were seeing was our own application fault, where a module was forcing all attributes to be loaded upon initialization.

**Pull request closed without merging**